### ISSUE TITLE
Enhance MAME HLSL To Do Sub-Refresh Processing On High-Hz Displays (CRT emulation, Reduces Blur, Lowers Lag, De-Stuttering)

### ISSUE THREAD

**Issue by mdrejhon** [mamedev/mame#6762]
# Here's the short TL;DR summary:

### **TL;DR: Enhance existing MAME HLSL to include sub-refresh behaviors, including simulating CRT phosphor fade in sub-refresh increments + optional rolling-scan BFI in sub-refresh increments.  All achieved via software, with the only hardware requirement being a high-Hz display, even 120Hz or 240Hz.  Made possible simply by a display having a higher Hz than the original emulated Hz**

***Bonus optional side effect 1: Reduce input lag.**  Since it's a rolling scan emulation via subdividing emulator Hz over multiple destination Hz, there is a simple cross-platform beam-race opportunity to reduce MAME lag even closer to 1:1 symmetry to original machine.  For example, 240Hz would use 4 hardware refresh cycles to emulate a 60Hz emulator refresh cycle in quarters at a time.  No raster knowledge of the destination display is needed, you're simply using the finer granularity of the higher output Hz.*

***Bonus optional side effect 2: Real-time standards conversion.** Smooth 50fps on 60Hz displays! The same algorithm I propose, accidentally de-stutters frame rate conversions too, so can be used for standards conversion (e.g. [ANIMATION: Software algorithm that makes 50fps look smooth without stutters on a 60hz non-VRR display](https://www.testufo.com/framerates-versus#photo=dota2-bg.jpg&pps=1920&framepacingerror=0&direction=rtl&framerate=50&compare=1&showfps=1)). So this algorithm is accidentally useful on today's 60Hz displays too. Output Hz actually can also be lower than emulator Hz as the algorithm can also blend refresh cycles together as part of the flexible Hz-agnostic phosphor fade algorithm capable of overlapping refresh cycles, similiar to what I successfully did in the link above. Allowing smooth 60fps on 50Hz displays too!*

# Long Version Below

MAME HLSL revolutionized the spatial look of a CRT tube.

However, it doesn't yet look temporally like a CRT tube (zero blur, etc).

The vision is that with increasingly-higher monitor refresh rates, thanks to the arrival of **240Hz IPS** (ASUS) and **360Hz IPS** (DELL AW2521H) this summer, now it is time to start talking about **emulating a CRT gun at the sub-refresh-cycle level**.

I posted at GroovyMAME about the concept of "Temporal HLSL"
http://forum.arcadecontrols.com/index.php/topic,162926.0.html

The long-term futurist vision is that a 1000Hz display will be able to emulate a CRT electron gun, by displaying a "rolling bar" with a phosphor fade trail.   Instead, we aim to do that in software, ala a "Temporal HLSL" algorithm of a rolling-bar software BFI.

**This may be easier to imagine:**
Much like playing back a 1000fps high-speed video of a CRT (with full dynamic range, not over exposed), back to a real-time 1000Hz sample-and-hold display.  The rolling bar.  Early tests show it ends up looking like the original CRT temporally with low persistence (1ms persistence).  Now that experimential ultra-Hz displays are in the lab.  360Hz is the beginnings of the "wow" factor in software-based rolling-bar BFI algorithms.  Finer-granularity Hz is even better, but 360Hz begins to start the "wow" in temporal look-n-feel.

ASUS has confirmed a long-term road map to 1000Hz displays (those unaware, can also read more; [Blur Busters Law: The Amazing Journey To Future 1000Hz Displays](https://blurbusters.com/1000hz-journey), with scientific citations included), and with the push of 120Hz becoming more mainstream (iPhone/Galaxy), high-Hz is expected to be inexpensive inclusions in future panels in ever-increasing numbers with refresh rates doubling approximately 5-10 years.  

While crazy numbers, the impetus of ultra-Hz is currently esports and VR.  The need for screens to emulate real life (real life doesn't flicker, which requires blurless sample-and-hold.  And blurless sample-and-hold requires insane Hz), thanks to VR and Holodecks, but a **side effect of the refresh rate race is an improved ability to emulate a CRT electron gun** at subrefresh timescales (zero-blur rolling scan), which is a boon for emulator originality.

Refresh rates are now getting high enough (360Hz) to make **Temporal HLSL** practical.  At 360Hz, one can use a 1/6th screen-height "rolling phosphor bar emulation" (with a gamma-corrected alpha-blended phosphor trail), generating only 2.7ms of motion blur (1/360sec persistence).  Bar height and bar fadetrails (as the bar scrolls downwards) can be programmable with existing options or new options in the configuration file.  One would have have alphablend-overlap betwen adjacent refresh cycles so you don't have artifacts, but done correctly, it looks seamless once we're in the refresh rate stratospheres.

In addition, I also posted a **permanent universal solution for software-BFI burn-in artifacts.** in [that thread](http://forum.arcadecontrols.com/index.php/topic,162926.0.html).  It is brighter, more colorful, no chessboard artifact, no banding or color depth loss, no burn in.

I'd like to see long-term discussion on incubating Temporal HLSL in what I envision is an approximately 5-year time window.  I don't expect this to be added to MAME In near term but I would like to see the community do longterm-planing for sub-refresh processing.   

CRTs will not last forever -- the factories have stopped.  We will need to emulate CRTs at sub-refresh timescales.   It has to happen within 10 years, when arcade tubes become valuable antiques.  

Thankfully the refresh rate race has made that feasible for originality, the dream of Year 2030s bringing 1000Hz OLEDs and MicroLEDs replicating the motion blur of a CRT rolling scan.   But we can get prepared now, with this year's 240Hz IPS and 360Hz IPS panels.

_P.S. I am open to moving this discussion to a different forum, or keeping this as an incubator github thread.  If this is the wrong place, open discussion on how to begin a "Temporal HLSL" collaborative initiative somewhere else is welcome.  Let's either (A) continue this issue as incubate this concept, or (B) figure out appropriate venue before closing this issue.  As MAME HLSL is incredibly well known for its studious attention to spatial detail, ultra-detailed configuration files, so it is the Gold Standard of CRT filters.  And the discussion to eventually make it capable of subrefresh processing is desirable to duplicate the original temporal behaviours of a CRT (including zero CRT motion blur).   Which ones of you are responsible for HLSL?_

**Comment by mdrejhon** [comment]
BTW, this is a very cross-platform idea.  This is much easier to conceptualize than beamraced VSYNC (while not needing to be mutually exclusive).

## Can be beam raced without knowing platform-specific or display-specific behaviors

Basically sub-refresh latencies using ordinary VSYNC ON.   You wouldn't have to care about how a display refreshes.  For example, at 240Hz, you'd display 1/4th of an emulator refresh cycle (rasterplotted in real time), in 4 separate output frames.  

**For a 60Hz emulator module onto a 240Hz monitor**
* Emulate 1/240sec, output 1/4-height frame at top
* Emulate 1/240sec, output 1/4-height frame just above center
* Emulate 1/240sec, output 1/4-height frame just below center
* Emulate 1/240sec, output 1/4-height frame at bottom

The rest of the frame would be black.  

Make top/bottom edges of bars fuzzy to reduce/eliminate tearing artifacts (avoid emulating the look of VSYNC OFF tearing). I have confirmed that you have to overlap the bars between output refresh cycles.  Use alpha-blend-to-black slightly beyond the 1/4-height frames for the 240Hz situation example.  This will prevent tearing artifacts. Make sure to gamma-correct the alphablend overlaps. Remember that RGB(128,128,128) is not exactly half the photons of RGB(255,255,255).  Use a configurable gamma correction number in HLSL config file.

## It can also acomplish beam raced latency without black frames too

The rolling scan BFI could in theory be configurable to full persistence.  It would achieve beam raced latencies to the output refresh cycle granularity.

**For a 60Hz emulator module onto a 240Hz monitor**
* Emulate 1/240sec, output frame with top 1/4 new refresh, bottom 3/4 old refresh
* Emulate 1/240sec, output frame with top 1/2 new refresh, bottom 1/2 old refresh
* Emulate 1/240sec, output frame with top 3/4 new refresh, bottom 1/4 old refresh
* Emulate 1/240sec, output frame with whole screen new refresh

Make sure to alphablend the seams (blur the refresh overlap line) to prevent tearing artifacts, otherwise it looks like "a VSYNC OFF emulation in VSYNC ON".  The alphablend fixes the tearing.  

This allows beam-raced latencies via sheer brute Hz.  So, this module could be configurable to have full persistence or rolling low persistence (rolling black period).

This will scale very well to the future, in the refresh rate race to retina refresh rates too.  A future 1000Hz display would emulate original machine latency to an error margin of 1ms, duplicating sub-refresh original-machine latencies, regardless of display technology (scan direction, display refresh pattern, etc).   Most high-Hz gaming displays have a sub-refresh processing delay, so the higher Hz you go, the more it converges into original-machine latency.  

Very little crossplatform dependencies, the only thing needed is VSYNC ON and the ability to do framerate=Hz.  

## Emulator Hz and destination Hz doesn't need to be divisible.

**Situation Example of 60Hz CRT emulation onto a 200Hz LCD** ... This formula is very Hz-agnostic.

Emulator Refresh Cycle #1:
....Real Refresh 1: full 60/200th height bar (30% screen height), at 0%-30% vertical position
....Real Refresh 2: full 60/200th height bar (30% screen height), at 30%-60% vertical position
....Real Refresh 3: fuill 60/200th height bar (30% screen height), at 60%-90% vertical position
....Real Refresh 4: 1/3 of 60/200th height bar (10% screen height), at 90%-100% vertical position

Emulator Refresh Cycle #2:
....Real Refresh 5: 2/3 of 60/200th height bar (20% screen height), at 0%-20% vertical position
....Real Refresh 6: full 60/200th height bar (30% screen height), at 20%-50% vertical position
....Real Refresh 7: full 60/200th height bar (30% screen height), at 50%-80% vertical position
....Real Refresh 8: 2/3 of of 60/200th height bar (20% screen height), at 80%-100% vertical position

Emulator Refresh Cycle #3:
....Real Refresh 9: 1/3 of 60/200th height bar (10% screen height), at 0%-10% vertical position
....Real Refresh 10: full 60/200th height bar (30% screen height), at 10%-40% vertical position
....Real Refresh 11: full 60/200th height bar (30% screen height), at 40%-70% vertical position
....Real Refresh 12: full 60/200th height bar (30% screen height), at 70%-100% vertical position

This is just a conceptual example of temporal compensation.  Algorithmically, this can be used for black frame insertion (rolling bar, some bars with image data, other bars black, with alphablended bleed overlap) or for "beam racing via brute Hz" (all bars with image data, new emu Hz overwriting old emu Hz, alphablend/blur the seams) or both simultaneously (rolling BFI + beam racing simultaneously).   You could adjust the alphablend factor up/down.

## Theoretically can also acomplish stutterless standards conversion (50Hz onto 60Hz)

Using a SIMILAR formula above, combined with a huge overlap alphablend adjustment (i.e. 50% or 75% screen height of overlap, perhaps).  Basically a scanout-enhanced version of a common alphablend standards-conversion algorithm.   It would thus, eliminate stutters.  The taller the alphablend overlap, the less stutter of standards-conversion.  The bar-overlap should be a configurable parameter in the configuration file.

This would conceptually be a more advanced version of the software-based variable refresh animation: www.testufo.com/vrr ....where I'm able to play any framerate stutterlessly onto any refreshrate, with a very simple alphablend algorithm not too dissimilar from the common 50/60 standards conversion alphablend algorithm.

So we're just "abusing" the alphablend overlap feature of this theoretical "Temporal HLSL" as the method of destutter in much the same stutter-eliminating way.

## Temporal HLSL concept is a universal CRT emulator with amazing crossplatform flexibility

It scale up and down universally

- Can temporally emulate a CRT electron gun at the granularity of destnation Hz
- Map any Hz to any Hz without stutter
- Benefits for stutterles Hz-conversion (50Hz onto 60Hz) 
- Benefits for high Hz (i.e. 60Hz onto 360Hz) 
- Potentially beam raceable (Temporal HLSL = a virtualized software display)
- Flickerless option (full persistence option)
- Flicker option (simulate CRT phosphor, including decay, subject to output-Hz granularity)

## OPTIONAL: Theoretically compatible with optional beamraced VSYNC approaches

<Optional>
- Temporal HLSL could emulate a higher-Hz display internally independently of output Hz
- A flywheel sync algorithm can optionally kick in whenever emuHz = outputHz, allowing a separate module to frameslice-beamrace this virtualized internal display to a low-Hz output display.

Thus, it would produce latencies identical to the Lagless VSYNC / Beamraced VSYNC approaches, like the one already implemented in WinUAE https://github.com/tonioni/WinUAE/issues/133 

Certainly, actual-hardware beam racing would be 100% optional (since it requires minor platform-dependant and display-scanout-direction knowledge), but I only simply add this to say that Temporal HLSL is 100% compatible simply by internally virtualizing a higher-Hz display internally, e.g. doing 600 frames per second of Temporal HLSL into RAM, and then doing 600-frameslice beamracing onto a real 60Hz display, subject to jitter margin to hide problems caused by CRT-curvature algorithms as well as computer performance jitter.
</Optional>

**Comment by mdrejhon** [comment]
Some information, @TomHarte already did some rolling-scan emulation (full-persistence version AFAIK) in the CRT emulator of CLK.   Though I am not sure of its current suitability for handling MAME Temporal HLSL.

https://github.com/TomHarte/CLK

Being founder of Blur Busters / TestUFO, I also have spare 240Hz monitors I may be able to loan/give out (North America only -- shipping is epensive) for a reputable developer fully committed to adding a high-Hz-aware CRT electron gun emulator.  240Hz is the entry level minimum where software based beam scanning becomes practical.

**Comment by mdrejhon** [comment]
Crossposting a potential "MAME Temporal HLSL" pre-requisite here from a RetroArch pull request, but this is such a simple universal algorithm.  The original text at https://github.com/libretro/RetroArch/pull/11342 but it is a generic algorithm applicable to MAME.   This might need to be 

> This is a universal generic crossplatform algorithm that should eventually become a best-practice for emulators in the next ten years.
> 
> # Problem: Existing frame pacing algorithm not future proof enough
> 
> Currently, G-SYNC uses a software-based frame pacing but it is not currently optimized in a future-proof way yet:
> 
> - Problem #1: Present imprecision create visible effects (stutters during VRR, flicker during BFI) especially in CPU heavy scenarios (slower systems, intensive emulators, big RunAhead settings)
> - Problem #2: Present imprecision makes it harder to add support for future enhancements (better VRR, >120Hz BFI (#10754), 
> combine VRR+BFI (#10754), beam racing (#6984), rolling-scan BFI CRT emulators (#10757)
> 
> # The upgrade to existing frame pacing algorithm
> 
> I propose a **separate thread responsible for frame presenting** (e.g. Present() or glxxSwapBuffers() or whichever API) that does the following:
> 
> - **Allows frame presents to continue independently of rendering**
> Presents will have no jitter even if emulator modules use darn near 100% CPU
> - **Allows frame presents to be optionally ultra precise**
> This can be done via busywait instead (or in addition to) of timer events -- because some algorithms (beamrace or BFI+VRR) have visibile artifacts with sub-millisecond errors.  Also, one can also timer-event to 0.5ms prior, then busywait rest of way.
> - **Allows same frame pacing algorithms to work with all sync technologies**
> (VSYNC ON, VSYNC OFF, DWM, triple buffer, AMD Enhanced Sync, NVIDIA Fast Sync, FreeSync, G-SYNC, VESA Adaptive-Sync, BFI, etc), making it much easier to combine them (e.g. BFI during VRR)
> - **Allows easier future addition of new algorithms**
> (e.g. rolling BFI CRT emulators, or lagless VSYNC beam racing), with little or no modification to emulator rendering
> - **More future proof**
> 
> # Largely a Streamlining of Existing Workflow
> 
> The existing present call would be replaced by a wrapper that passes the frame to a frame presentation thread.  The presentation thread will time the presentation itself.
> 
> Some of the workflow already exists, it just needs to be re-jigged into an official unified workflow with capability of improved precision.
> 
> - "Rendering Thread" = Thread that runs the emulator and generates the emulator frames;
> - "Presenting Thread" = Thread that is now permanently responsible for frame presentation;
> - "Present Wrapper" = This replaces the existing frame present method (e.g. glxxSwapBuffers() or Present() or whatever platform API is used to pass frame to the graphics drivers).  So that *Rendering Thread* can transfer (or copy) frame to the *Presenting Thread*
> 
> # Suggested Stage 1 Workflow
> 
> 1. *Presenting Thread* only purpose is presenting frames (no rendering)
> 2. *Presenting Thread* is always higher priority than rendering thread, for purpose of timing precision.  Most of the time, presenting thread uses 0% CPU since it's just timing pre-rendered frames, so high precision becomes harmless to *Rendering Thread*
> 3. *Presenting Thread* can optionally be forced to present immediately so it is backwards compatible with existing present workflow (this can ease iterative development) or for platforms not stable with separate-thread presenting (Fortunately, I don't think there's are any left).
> 4. *Rendering Thread* should do all rendering, including CRT filters
> 5. *Present Wrapper* inside the *Rendering Thread* can be a good way to hide/centralize all the final processing.  Such as adding CRT filters, or rendering a whole sequence of BFI framebuffers (low emulator Hz on high real-display Hz).   This hides the implementation details of many refreshing algorithms, and makes cross-platform easier.
> 6. *Presenting Thread* will make sure that the time intervals between consecutive presents are as exact as possible.   When this is achieved, the algorithm suddenly become universal (works with all sync technologies).
> 7. *Present Wrapper* can still emulate the behavior of a 60Hz VSYNC ON waitable swapchain (regardless of whether underlying hardware is doing VSYNC ON or VSYNC OFF or VRR or BFI or whatever) by waiting for a heartbeat from the *Present Thread* 
> 
> Metaphorically, this workflow is a metaphorically **software-based VSYNC ON emulator**, hiding the quirks of GPU drivers or destination displays away from emulator rendering.  While **simultaneously improving user-friendliness** (things just works automatically upon startup) and **making things less buggy** (no VRR stutters, no BFI flicker) and **future proofing** (even BFI made VRR compatible).
> 
> In a 60Hz VSYNC ON scenario, this is just defacto passthrough behavior (*Present Thread* will immediately present), while allowing one framepacing algorithm to work with ALL sync technologies more reliably.   And it adds no extra workflow lag.
> 
> Don't worry about BFI for now (#10754 and/or #10757), don't worry about beamracing for now (#6984 and/or #10757); those are solvable in future (e.g. wrappers for PresentScanLine() can be added later to pass one pixel row between *Rendering Thread* to the *Present Thread*, as an example).  Just focus on this simple workflow.
> 
> # Easy Debugging Tip for non-VRR 60Hz Developers 
> 
> Testing without VRR can be done via 60Hz VSYNC OFF while using CPU-heavy emulation/emulation settings.  Use 60Hz VSYNC OFF, and use tearline jitter as a timing-precision debugger.  If the tearline erratically moves or jitters/vibrates massively, your present timing is not microsecond-accurate.  If the tearline is stationary or rolls slowly up/down, your present timing is near microsecond-accurate.   
> 
> 1080p 60Hz is a horizontal scanrate of 67 kilohertz (approx 67000 scanlines per second, including VBI).  So a 1/67000th second delay moves a VSYNC OFF tearline downwards by 1 pixel.   Modern displays still scan from top-to-bottom ([high speed videos](https://blurbusters.com/scanout)) and VSYNC OFF tearlines are a raster artifact.   
> 
> So if your tearline is vibrating by 50 pixels up/down, that means you've got a 50/67000th second imprecision in your Present() or glxxSwapBuffers() timing.   Thusly, VSYNC OFF 60Hz is an excellent timing debugger, since VSYNC OFF tearline is a real-display raster where the new real GPU framebuffer splices into the destination display's scanout position.
> 
> If you have a high-Hz display, you can also test 60fps at VSYNC OFF 120Hz or VSYNC OFF 240Hz for more sensitive timing-precision debugging (1/135000th second for a 120Hz tearline moving downwards by 1 pixel, for example)
> 
> If you succeed, then it help guarantees VRR users get amazing framepacing, an BFI users get flicker-free operation (even if you never test VRR or BFI)  **Thus, use 60 Hz VSYNC OFF as a clever easy debugger for frame-present timing precision if you don't have 144Hz or VRR or BFI!**

**Comment by mdrejhon** [comment]
--interesting commentary-- 

One conceptual way to more easily understand this is high speed video of a CRT tube.  You see a rolling bar in those, with blurry edges.  This is seen as an example on YouTube.

Did you know that playback of a 1000fps high-speed video of a CRT tube -- in real time onto a true 1000Hz display -- allows a 1000Hz display to perceptually emulate the original CRT?   

The "Temporal HLSL" concept, aims to emulate that behavior in software;

**Comment by mdrejhon** [comment]
I have received word that successful tests were done at lower frame rates (240fps studio-quality camera videos of a CRT played back on a true-240Hz LCD) and it scales pretty remarkably well!   

The higher the frame rate (and refresh rate), the more accurately the digital flat panel temporally emulates a CRT, proving the realisticness of the Temporal HLSL move.

One caveat is that the black periods does diminish lumens, but we found HDR can be a workaround to emulate the bright CRT electron beam, as long as there's sufficient window to capture (e.g. if max-HDR 10% window is supported, then a 16-slice 1000fps 1000Hz will be within the under-10% window HDR).   So HDR engineering (future HDR high-Hz displays) can help helps bring the necessary lumens surge to compensate for the blackness.

**Comment by angelosa** [comment]
> P.S. I am open to moving this discussion to a different forum, or keeping this as an incubator github thread.

First off: there's no such thing as an _incubator github thread_. This isn't a forum but a _github issue_ where **you [the user] want something out of MAME so that it benefit [to yourself and the MAME community]**.

Then I have a question about all this roadmap/mailinglist-like TL;DR: how much all this technical mumbo jumbo applies to MAME at **feature request** level? I read stuff that isn't even released atm (Asus 1000Hz monitors) or stuff that isn't emulated or interfaced to (any flavor of GoPro). Is there anything that can be pinned to your original first message as an **index of brief feature requests** in order to achieve this goal?

For now I'm gonna spike this out, as current things stands I'm not even sure it belongs to bgfx or ui/ux or just wontfix/invalid because it belongs to a very specific HLSL configuration that definitely won't fit in default scenario to the vast majority of users until 2060 or thereabouts.

**Comment by mdrejhon** [comment]
> > P.S. I am open to moving this discussion to a different forum, or keeping this as an incubator github thread.
> 
> First off: there's no such thing as an _incubator github thread_. This isn't a forum but a _github issue_ where **you [the user] want something out of MAME so that it benefit [to yourself and the MAME community]**.

Understood.  I'll edit the top with a TL;DR this year.  

Keep this issue around because we're under active discussion in other communities, since it likely will be incubated elsewhere before it arrives here.

BTW, 1000Hz is just an end goal.  The reality is **this MAME HLSL algorithm begins to be useful beginning at ~240Hz**, and those are getting more widespread.  And there is news that Apple iPhone is coming out with a 240Hz iPhone, since 240Hz reduces scrolling motion blur.  120Hz is quickly becoming semi-mainstream (almost all 4K HDTVs support 120Hz, all new gaming consoles support 120Hz, new smartphones will soon come 120Hz standard, and DELL/HP is considering doubling Hz for office monitors for scrolling-ergonomic reasons).  After that, 240Hz is the next mainstream step.  **This will definitely happen before 2060**.

**Comment by mdrejhon** [comment]
I have renamed this github to the more self-explanatory name:

# "Enhance MAME HLSL To Do Sub-Refresh Processing On High-Hz Displays (CRT emulation, Reduces Blur, Lowers Lag, De-Stuttering)"

**Comment by stilett0** [comment]
This isn't a Pull Request. You used ten thousand words when ten would do. If it's a Pull Request, then there needs to be Code. Where's your code, @mdrejhon ? You don't have code, you at best have an algorithm. But really all you have is ten thousand words running off at the mouth. Start writing code, stop with the theorizing. It all sounds quasi-scientific, sure, but it's all a smokescreen to try to instigate someone else to do the actual work. It's all conceptual, nothing practical. And the fact that you haven't actually contributed written code here in this PR makes me suspect that you can't.

Between THAT, and the fact that some folks on Twitter conflated this "Pull Request" as something that MAMEdev as a team themselves had "officially" added as a project roadmap bulletpoint... (January: https://twitter.com/_daemons/status/1351450119112048640 ) I'm closing this. You don't get to slide your Feature Request agenda items in through the back door that way. Frankly, it's unwanted attention-seeking behavior.

This isn't a Pull Request, this is a Feature Request. At best, it belongs in Issues, not Pull Requests.

And I don't look kindly upon a Feature Request to try to pre-emptively add support for hardware, that practically doesn't exist yet, and that won't be mainstream absorption for quite some years to come. This does nothing for us. You're just being Master of the Obvious here. We know there will be gamechanging high-refresh displays coming out or out already, we don't need reminders.

**Comment by happppp** [comment]
@stilett0 wrongly claims this is a Pull Request. Maybe he can clarify later.

IMO this issue can remain closed for now. ultra-Hz displays don't exist yet (fot the consumer)
Likewise, we don't have an Issue about quantum computers enhancing accuracy for emulated netlist games.

**Comment by mdrejhon** [comment]
In all friendliness, 

1. This is a feature request, not a pull request.

2. True 120Hz and 240Hz displays exist already, and there is already a 360 Hz display.  I simply extrapolated this to far future.   

I would like to re-post the TL;DR:

Here's the short TL;DR summary:
TL;DR: Enhance existing MAME HLSL to include sub-refresh behaviors, including simulating CRT phosphor fade in sub-refresh increments + optional rolling-scan BFI in sub-refresh increments. All achieved via software, with the only hardware requirement being a high-Hz display, even 120Hz or 240Hz. Made possible simply by a display having a higher Hz than the original emulated Hz
Bonus optional side effect 1: Reduce input lag. Since it's a rolling scan emulation via subdividing emulator Hz over multiple destination Hz, there is a simple cross-platform beam-race opportunity to reduce MAME lag even closer to 1:1 symmetry to original machine. For example, 240Hz would use 4 hardware refresh cycles to emulate a 60Hz emulator refresh cycle in quarters at a time. No raster knowledge of the destination display is needed, you're simply using the finer granularity of the higher output Hz.

Bonus optional side effect 2: Real-time standards conversion. Smooth 50fps on 60Hz displays! The same algorithm I propose, accidentally de-stutters frame rate conversions too, so can be used for standards conversion (e.g. ANIMATION: Software algorithm that makes 50fps look smooth without stutters on a 60hz non-VRR display). So this algorithm is accidentally useful on today's 60Hz displays too. Output Hz actually can also be lower than emulator Hz as the algorithm can also blend refresh cycles together as part of the flexible Hz-agnostic phosphor fade algorithm capable of overlapping refresh cycles, similiar to what I successfully did in the link above. Allowing smooth 60fps on 50Hz displays too!

**Comment by cuavas** [comment]
>This is a feature request, not a pull request.

It’s a pie-in-the-sky dream.
